types/generate.js

```diff
@@ -0,0 +1,114 @@
+// TypeScript definitions are generated here.
+// AVA allows chaining of function names, like `test.after.cb.always`.
+// The order of these names is not important.
+// Writing these definitions by hand is hard. Because of chaining,
+// the number of combinations grows fast (2^n). To reduce this number,
+// illegal combinations are filtered out in `verify`.
+// The order of the options is not important. We could generate full
+// definitions for each possible order, but that would give a very big
+// output. Instead, we write an alias for different orders. For instance,
+// `after.cb` is fully written, and `cb.after` is emitted as an alias
+// using `typeof after.cb`.
+
+var path = require('path');
+var fs = require('fs');
+
+var base = fs.readFileSync(path.join(__dirname, 'base.d.ts')).toString();
+
+// All suported function names
+var allParts = ['serial', 'before', 'after', 'skip', 'todo', 'failing', 'only', 'beforeEach', 'afterEach', 'cb', 'always'];
+
+var output = base + generatePrefixed([]);
+fs.writeFileSync(path.join(__dirname, 'generated.d.ts'), output);
+
+// Generates type definitions, for the specified prefix
+// The prefix is an array of function names
+function generatePrefixed(prefix) {
+	var output = 'export namespace ' + ['test'].concat(prefix).join('.') + ' {\n';
+	
+	var children = '';
+	var empty = true;
+	for (var i = 0; i < allParts.length; i++) {
+		var part = allParts[i];
+		var parts = prefix.concat([part]);
+		
+		if (prefix.indexOf(part) !== -1 || !verify(parts, true)) {
+			// Function already in prefix or not allowed here
+			continue;
+		}
+		
+		// Check that `part` is a valid function name.
+		// `always` is a valid prefix, for instance of `always.after`,
+		// but not a valid function name.
+		if (verify(parts, false)) {
+			empty = false;
+			if (!isSorted(parts)) {
+				output += '\t' + 'export const ' + part + ': typeof test.' + parts.sort().join('.') + ';\n';
+				continue;
+			} else if (parts.indexOf('todo') !== -1) {
+				output += '\t' + writeFunction(part, 'name: string', 'void');
+			} else {
+				var type = testType(parts);
+				output += '\t' + writeFunction(part, 'name: string, implementation: ' + type);
+				output += '\t' + writeFunction(part, 'implementation: ' + type);
+			}
+		}
+		
+		children += generatePrefixed(parts);
+	}
+	if (empty) {
+		output = '';
+	} else {
+		output += '}\n';
+	}
+	return output + children;
+}
+function writeFunction(name, args) {
+	return 'export function ' + name + '(' + args + '): void;\n';
+}
+
+function verify(parts, asPrefix) {
+	var has = arrayHas(parts);
+	if (has('only') + has('skip') + has('todo') > 1) return false;
```

Comment:
Although it works, it feels kinda hacky. I think I'd rather prefer something like

```js
has('only', 'skip', 'todo').filter(Boolean).length
```

Or something like that.